src/core/adt/serialize.js

```diff
@@ -0,0 +1,32 @@
+const { tagSymbol, typeSymbol } = require('./core');
+const mapValues = require('folktale/core/object/map-values');
+
+const assertType = (given, expected) => {
+  if (expected !== given) {
+    throw new TypeError(`
+       The JSON structure was generated from ${expected}.
+       You are trying to serialize it into ${given}. 
+    `);
+  }
+};
+const parseValue = (parsers) => (value) =>
+  parsers[value.typeName] ? parsers[value.typeName].fromJSON(value, parsers)
```

Comment:
This needs to check for `null` and `undefined`, since JS throws errors when you try to treat those as objects. This could be `value == null ? value : parsers[value.typeName] ? ...`